src/lib/mixer/mixer_multirotor.py

```diff
@@ -0,0 +1,380 @@
+Description: This script can be used to prototype new mixer algorithms and test
+it against the C++ implementation.
+"""
+
+
+from __future__ import print_function
+
+from argparse import ArgumentParser
+import numpy as np
+import numpy.matlib
+import subprocess
+
+
+# --------------------------------------------------
+# mixing algorithms
+# --------------------------------------------------
+
+def compute_desaturation_gain(u, u_min, u_max, delta_u):
+    """
+    Computes the gain k by which delta_u has to be multiplied
+    in order to unsaturate the output that has the greatest saturation
+    """
+    d_u_sat_plus = u_max - u
+    d_u_sat_minus = u_min - u
+    k = np.zeros(u.size*2)
+    for i in range(u.size):
+        if abs(delta_u[i]) < 0.000001:
+            # avoid division by zero
+            continue
+
+        if d_u_sat_minus[i] > 0.0:
+            k[2*i] = d_u_sat_minus[i] / delta_u[i]
+        if d_u_sat_plus[i] < 0.0:
+            k[2*i+1] = d_u_sat_plus[i] / delta_u[i]
+
+    k_min = min(k)
+    k_max = max(k)
+
+    # Reduce the saturation as much as possible
+    k = k_min + k_max
+    return k
+
+
+def minimize_sat(u, u_min, u_max, delta_u):
+    """
+    Minimize the saturation of the actuators by
+    adding or substracting a fraction of delta_u.
+    Delta_u is the vector that added to the output u,
+    modifies the thrust or angular acceleration on a
+    specific axis.
+    For example, if delta_u is given
+    to slide along the vertical thrust axis, the saturation will
+    be minimized by shifting the vertical thrust setpoint,
+    without changing the roll/pitch/yaw accelerations.
+    """
+    k_1 = compute_desaturation_gain(u, u_min, u_max, delta_u)
+    u_1 = u + k_1 * delta_u # Try to unsaturate
+
+
+    # Compute the desaturation gain again based on the updated outputs.
+    # In most cases it will be zero. It won't be if max(outputs) - min(outputs)
+    # > max_output - min_output.
+    # In that case adding 0.5 of the gain will equilibrate saturations.
+    k_2 = compute_desaturation_gain(u_1, u_min, u_max, delta_u)
+
+    k_opt = k_1 + 0.5 * k_2
+
+    u_prime = u + k_opt * delta_u
+    return u_prime
+
+def mix_yaw(m_sp, u, P, u_min, u_max):
+    """
+    Mix yaw by adding it to an existing output vector u
+
+    Desaturation behavior: thrust is allowed to be decreased up to 15% in order to allow
+    some yaw control on the upper end. On the lower end thrust will never be increased,
+    but yaw is decreased as much as required.
+    """
+    m_sp_yaw_only = np.matlib.zeros(m_sp.size).T
+    m_sp_yaw_only[2, 0] = m_sp[2, 0]
+    u_p = u + P * m_sp_yaw_only
+
+    # Change yaw acceleration to unsaturate the outputs if needed (do not change roll/pitch),
+    # and allow some yaw response at maximum thrust
+    u_r_dot = P[:,2]
+    u_pp = minimize_sat(u_p, u_min, u_max+0.15, u_r_dot)
+    u_T = P[:, 3]
+    u_ppp = minimize_sat(u_pp, -1000, u_max, u_T)
```

Comment:
use `u_min` instead of -1000. Should we add the flag `reduce only` like in the cpp code?